### PR TITLE
CI architecture and usage documentation

### PR DESCRIPTION
**Pull request by quazirafi**
Added complete documentation of the CI architecture and usage of reusable CI components.

### CODE REVIEW

**Review by ryanmrichard — APPROVED** [review]
LGTM 

@twindus more of a management issue, but I think we should probably decide on a long term CI maintainer. In particular, I'm thinking there should probably be a staff scientist who is in charge of CI. This person doesn't necessarily need to do the CI work, but they need to be aware of what's going on and generally speaking how the CI works. My point is whoever this person is should probably review this PR and make sure it makes sense to them (yes I'm hoping to not nominate myself :wink:). At one point this was Huub (and Kris). So @hjjvandam are you willing to volunteer?

File: docs/source/project_management/ci_maintainer_notes.rst [hunk 1]
```diff
@@ -7,97 +7,207 @@ The purpose of this page is to document various aspects of NWX's CI.
...
+4. LAYER OF BASH SCRIPTS (BOTTOM LAYER)
+---------------------------------------
+
+**Bash scripts of .github repository:** The bottom layer consists of the bash scripts that are used by the composite actions in the previous layer. The bash scripts are also developed in a generic manner so that they can be re-used across multiple repository under the NWX organization. These bash scripts contain the bash commands necessary for building, testing and deploying the application and the corresponding documentation. The bash scripts resides in the individual action folders of the ``.github`` repository.
+
+The bash scripts are shown as ``Action{1,2,3}_bash_{1,2}.sh`` in the architecture figure.
+
+
+CI coverage
+=============================
```
[the remark this answers is not in the document]
Comment (quazirafi) [reply docs/source/project_management/ci_maintainer_notes.rst]: Done.

**Comment by twindus** [comment]
Hey @hjjvandam!  Is this something that you think you can take him.

**Comment by hjjvandam** [comment]
@ryanmrichard I can take this on. I just need to get back into it.

**Comment by ryanmrichard** [comment]
@hjjvandam sounds good. The CI team is still hard at work, so at this point we really just you to: read through these notes, point out anything that's not clear, and let us know if you have any design/architecture questions.

**Comment by hjjvandam** [comment]
In "ci_maintainer_notes.rst" there still are mentions of "test_docs.yaml" and "test_docs_tmpl.yml". I guess they should be "build_docs.yaml" and company?

**Comment by ryanmrichard** [comment]
@quazirafi I resolved a conflict, between this PR and master. You may want to take a quick look and make sure I resolved it correctly (basically we want your version and no part of the old one).

**Comment by quazirafi** [comment]
> @quazirafi I resolved a conflict, between this PR and master. You may want to take a quick look and make sure I resolved it correctly (basically we want your version and no part of the old one).

I think you went back to the older documentation.

**Comment by ryanmrichard** [comment]
@quazirafi [this](https://github.com/NWChemEx-Project/.github/blob/48307d5e96bc773c877eec776c40928c5f75729a/docs/source/project_management/ci_maintainer_notes.rst) is the state of the CI maintainer notes in the most recent commit to this PR. Quickly glancing through it, it looks like they are your notes (I don't see any of the old notes). Is this what you were looking at too?

**Comment by quazirafi** [comment]
> @quazirafi [this](https://github.com/NWChemEx-Project/.github/blob/48307d5e96bc773c877eec776c40928c5f75729a/docs/source/project_management/ci_maintainer_notes.rst) is the state of the CI maintainer notes in the most recent commit to this PR. Quickly glancing through it, it looks like they are your notes (I don't see any of the old notes). Is this what you were looking at too?

@ryanmrichard yes these are my notes. Actually I directly went to the documentation link. I guess that was not updated. So, I found the previous notes. Looks ok to me.

**Comment by twindus** [comment]
Thanks @hjjvandam!

**Comment by quazirafi** [comment]
> In "ci_maintainer_notes.rst" there still are mentions of "test_docs.yaml" and "test_docs_tmpl.yml". I guess they should be "build_docs.yaml" and company?

The repos that follow our architecture use deploy_docs.yaml and test_docs.yaml. And these workflows then call the callee workflows deploy_docs_tmpl.yml and test_docs_tmpl.yml of centralized .github repo. @hjjvandam

**Comment by hjjvandam** [comment]
> > In "ci_maintainer_notes.rst" there still are mentions of "test_docs.yaml" and "test_docs_tmpl.yml". I guess they should be "build_docs.yaml" and company?
> 
> The repos that follow our architecture use deploy_docs.yaml and test_docs.yaml. And these workflows then call the callee workflows deploy_docs_tmpl.yml and test_docs_tmpl.yml of centralized .github repo. @hjjvandam

OK, but this still is confusing. Under "LAYER OF CALLER WORKFLOWS (TOPMOST LAYER)" there are 4 YAML files listed. One of those is called "build_docs.yaml". But then a bit further down it says "d. ``test_docs.yaml`` calls ``test_docs_tmpl.yml`` template workflow with input ``target``, ``dependencies`` and ``skip_doxygen``." However, "test_docs.yaml" was never mentioned before. So is it possible that "build_docs.yaml" should really be "test_docs.yaml"?

**Comment by quazirafi** [comment]
> > > In "ci_maintainer_notes.rst" there still are mentions of "test_docs.yaml" and "test_docs_tmpl.yml". I guess they should be "build_docs.yaml" and company?
> > 
> > 
> > The repos that follow our architecture use deploy_docs.yaml and test_docs.yaml. And these workflows then call the callee workflows deploy_docs_tmpl.yml and test_docs_tmpl.yml of centralized .github repo. @hjjvandam
> 
> OK, but this still is confusing. Under "LAYER OF CALLER WORKFLOWS (TOPMOST LAYER)" there are 4 YAML files listed. One of those is called "build_docs.yaml". But then a bit further down it says "d. `test_docs.yaml` calls `test_docs_tmpl.yml` template workflow with input `target`, `dependencies` and `skip_doxygen`." However, "test_docs.yaml" was never mentioned before. So is it possible that "build_docs.yaml" should really be "test_docs.yaml"?

@hjjvandam You are absolutely right. That was an error from me. Thanks for pointing this out. I corrected it. @ryanmrichard Maybe you need to merge this change too!

**Review by ryanmrichard — APPROVED** [review]

**Pull request merged**